File: scriptcore/Source/Paper/Core/PaperID.cs
using System;

namespace Paper
{
    public class PaperID
    {
        private ulong _id;

        public PaperID(ulong _ID)
        {
            _id = _ID;
        }

        public PaperID(string _UUID)
        {
            _id = Convert.ToUInt64(_UUID, 16);
        }

        public PaperID(PaperID _PaperID)
        {
            this._id = _PaperID._id;
        }

        public void Set(ulong _UUID)
        {
            _id = _UUID;
        }

        public void Set(string _UUID)
        {
            _id = Convert.ToUInt64(_UUID, 16);
        }

        public override string ToString()
        {
            return _id.ToString("x16");
        }

        public ulong ToUlong()
        {
            return _id;
        }

        public static implicit operator ulong(PaperID _PaperID)
        {
            return _PaperID.ToUlong();
        }

        public static implicit operator PaperID(ulong _ID)
        {
            return new PaperID(_ID);
        }

        public static implicit operator string(PaperID _PaperID)
        {
            return _PaperID.ToString();
        }

        public static implicit operator PaperID(string _ID)
        {
            return new PaperID(_ID);
        }
    }
}

File: engine/src/core/utils/PaperID.cpp
#include "Engine.h"
#include "PaperID.h"

#include "scene/Entity.h"

namespace Paper
{
	static std::random_device randomDevice;
	static std::mt19937_64 engine(randomDevice());
	static std::uniform_int_distribution<uint64_t> distribution;

	PaperID::PaperID()
		: uuid(distribution(engine))
	{
	}

	PaperID::PaperID(uint64_t uuid)
		: uuid(uuid)
	{
	}

	PaperID::PaperID(std::string id)
	{
		std::stringstream ss;
		ss << std::hex << id;
		ss >> uuid;
	}

	void PaperID::Set(std::string id)
	{
		std::stringstream ss;
		ss << std::hex << id;
		ss >> uuid;
	}

	void PaperID::Set(uint64_t id)
	{
		uuid = id;
	}

	bool PaperID::Empty() const
	{
		return !uuid;
	}

	uint64_t PaperID::toUInt64() const
	{
		return operator uint64_t();
	}

	std::string PaperID::toString() const
	{
		return operator std::string();
	}

	Entity PaperID::ToEntity() const
	{
		if (Scene::GetActive())
			return Scene::GetActive()->GetEntity(*this);
		return {};
	}
}
